source/Plugins.cpp

```diff
@@ -38,14 +64,135 @@ namespace {
 				continue;
 
 			for(const DataNode &child : node)
-				if(child.Size() == 2)
+				if(child.Size() == 3)
 				{
+					lock_guard<mutex> guard(pluginsMutex);
 					auto *plugin = plugins.Get(child.Token(0));
 					plugin->enabled = child.Value(1);
 					plugin->currentState = child.Value(1);
+					plugin->version = child.Token(2);
 				}
 		}
 	}
+
+	bool oldNetworkActivity = false;
+
+	mutex activePluginsMutex;
+	set<string> activePlugins;
+
+	// The maximum size of a plugin in bytes, this will be 1 GB.
+	const int MAX_DOWNLOAD_SIZE = 1000000000;
+
+	// Copy an entry from one archive to the other
+	void CopyData(struct archive *ar, struct archive *aw)
+	{
+		int retVal;
+		const void *buff;
+		size_t size;
+		int64_t offset;
+
+		while(true)
+		{
+			retVal = archive_read_data_block(ar, &buff, &size, &offset);
+			if(retVal == ARCHIVE_EOF)
+				return;
+			if(retVal != ARCHIVE_OK)
+				return;
+			if(archive_write_data_block(aw, buff, size, offset) != ARCHIVE_OK)
+				return;
+		}
+	}
+
+	bool ExtractZIP(std::string &filename, const std::string &destination, const std::string &expectedName)
```

Comment:
Is there really no library we can use to extract zips?

Reply:
this is the library to extract zips?

Reply:
Yeah, but one would think that there are external libraries that support it as a standalone zip operation, instead of having to do a ton of manual work.

Reply:
~~welcome to c++~~
there probably are but libarchive allows for more control and has the better support, also even if it was a lib specifically for zips, they still require the same workflow

so libarchive is the more solid choice allowing even other formats to be extracted the same way.